In [119]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [120]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Overwriting APlusB.cpp


In [132]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

4 8
12

### Test input output files

In [122]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [133]:
%%writefile output1.txt
12

Overwriting output1.txt


In [124]:
%%writefile input2.txt
0 9

Overwriting input2.txt


In [125]:
%%writefile output2.txt
9

Overwriting output2.txt


In [134]:
!./APlusB < input1.txt

12

In [135]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [128]:
!grep -xvFf output1.txt output1.txt

In [129]:
!grep -xvFf output1.txt output2.txt

9


In [130]:
!grep -xvFf output2.txt output1.txt

12


In [131]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [136]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [137]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


In [138]:
!python sum.py

2 5
7


In [139]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [140]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [141]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Overwriting max_pairwise_product_full.cpp


In [142]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

5
1 2 3 4 5
20


$$n^2 \ \text{first != second  comparisons},  n^2 - n \ \ \text{numbers[first] * numbers[second] products},  O(n^2)$$

### Test 1: Small numbers

In [154]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [155]:
!./max_pairwise_product_full < input1.txt

90


In [156]:
%%writefile output1.txt
90

Overwriting output1.txt


In [157]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [158]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [159]:
%%writefile output2.txt
90

Overwriting output2.txt


In [160]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 in the input

In [161]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Overwriting input3.txt


In [162]:
200000*200000

40000000000

In [164]:
%%writefile output3.txt
40000000000

Overwriting output3.txt


In [165]:
!./max_pairwise_product_full < input3.txt

1345294336


In [179]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [167]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {1..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [168]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


Changing the variables to long long

In [169]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long_full.cpp


In [170]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [172]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [173]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [174]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



----------------------------------------------------**Test *3.1***-------------------------------------------------

In [183]:
%%writefile input4.txt
10
7 8 9554  5214  42  567 331 44  4 1

Overwriting input4.txt


In [177]:
9554*5214

49814556

In [180]:
%%writefile ctest4.txt
49814556

Writing ctest4.txt


In [176]:
!./max_pairwise_product_long_full < input4.txt

49814556


In [191]:
%%writefile output4.txt
49814556

Overwriting output4.txt


In [181]:
!./max_pairwise_product_long_full < input4.txt > ctest4.txt
!grep -xvFf output4.txt ctest4.txt

----------------------------------------------------Test 3.2-------------------------------------------------

In [198]:
%%writefile input5.txt
5
23  5 4455  6871  25

Overwriting input5.txt


In [185]:
4455*6871

30610305

In [197]:
%%writefile ctest5.txt
30610305

Overwriting ctest5.txt


In [187]:
!./max_pairwise_product_long_full < input5.txt

30610305


In [203]:
%%writefile output5.txt
30610305

Overwriting output5.txt


In [192]:
!./max_pairwise_product_long_full < input5.txt > ctest5.txt
!grep -xvFf output5.txt ctest5.txt

---------------------------------------------------Test 3.3-------------------------------------------------

In [196]:
%%writefile input6.txt
20
12  3 6 7 918 4 1 2 85  3 7 23  45  68  72  23  100 58  75  64

Writing input6.txt


In [195]:
918*100

91800

In [199]:
%%writefile ctest6.txt
91800

Writing ctest6.txt


In [200]:
!./max_pairwise_product_long_full < input6.txt

91800


In [202]:
%%writefile output6.txt
91800

Writing output6.txt


In [204]:
!./max_pairwise_product_long_full < input6.txt > ctest6.txt
!grep -xvFf output6.txt ctest6.txt

----------------------------------------------------Test 3.4-------------------------------------------------

In [212]:
%%writefile input7.txt
3
233356464  112745  4556658

Overwriting input7.txt


In [214]:
233356464*4556658

1063325598537312

In [215]:
%%writefile ctest7.txt
1063325598537312

Overwriting ctest7.txt


In [213]:
!./max_pairwise_product_long_full < input7.txt

1063325598537312


In [220]:
%%writefile output7.txt
1063325598537312

Writing output7.txt


In [221]:
!./max_pairwise_product_long_full < input7.txt > ctest7.txt
!grep -xvFf output7.txt ctest7.txt

---------------------------------------------------Test 3.5-------------------------------------------------

In [231]:
%%writefile input8.txt
4
10101  10001101  101 1

Overwriting input8.txt


In [224]:
10101*10001101

101021121201

In [225]:
%%writefile ctest8.txt
101021121201

Writing ctest8.txt


In [226]:
!./max_pairwise_product_long_full < input8.txt

101021121201


In [228]:
%%writefile output8.txt
101021121201

Writing output8.txt


In [230]:
!./max_pairwise_product_long_full < input8.txt > ctest8.txt
!grep -xvFf output8.txt ctest8.txt

---------------------------------------------------Test 3.6-------------------------------------------------

In [233]:
%%writefile input9.txt
7
2 3 5 7 4 9 1

Writing input9.txt


In [232]:
7*9

63

In [234]:
%%writefile ctest9.txt
63

Writing ctest9.txt


In [235]:
!./max_pairwise_product_long_full < input9.txt

63


In [236]:
%%writefile output9.txt
63

Writing output9.txt


In [237]:
!./max_pairwise_product_long_full < input9.txt > ctest9.txt
!grep -xvFf output9.txt ctest9.txt

---------------------------------------------------Test 3.7-------------------------------------------------

In [239]:
%%writefile input10.txt
12
2 3 5 7 4 9 1 55 44 56 78 98

Overwriting input10.txt


In [240]:
78*98

7644

In [242]:
%%writefile ctest10.txt
7644

Writing ctest10.txt


In [243]:
!./max_pairwise_product_long_full < input10.txt

7644


In [244]:
%%writefile output10.txt
7644

Writing output10.txt


In [245]:
!./max_pairwise_product_long_full < input10.txt > ctest10.txt
!grep -xvFf output10.txt ctest10.txt

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and inoput12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting Test4Write.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

cc1plus: fatal error: WriteTest.cpp: No such file or directory
compilation terminated.
/bin/bash: line 1: ./WriteTest: No such file or directory


In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Writing output12.txt


Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

40000000000

real	5m6.795s
user	4m59.238s
sys	0m0.217s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	5m4.108s
user	4m57.506s
sys	0m0.172s


##### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[1, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 4 + n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed

    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                //1+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                               //n**2 -n
                    max_product = aux_product;                               //[1,n**2/2 - n/2]  upper bound
                    execints++;

                }
        }
      }
    }
                                                                            // --------------
    return max_product;                                                     // 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //N inst exec by line executed

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }



    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound


    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]

Overwriting max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 9 + 5n + \left[5n^2+1, 11\frac{n^2}{2}  - \frac{n}{2}\right] \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 5(10) + 5(10)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 589 $$

$$T(10) \in [560, 604]  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

   1  2  3  4  5  6  7  8  9  10
1  x  2  3  4  5  6  7  8  9  10         9
2  2  x  6  8  10 12 14 16 18 20         5 14
3  3  6  x  12 15 18 21 24 27 30         4 18
4  4  8  12 x  20 24 28 32 36 40         3 21
5  5  10 15 20 x  30 35 40 45 50         2 23
6  6  12 18 24 30 x  42 48 54 60         2 25
7  7  14 21 28 35 42 x  56 63 70         2 27
8  8  18 27 36 45 54 63 72 x  90         1 30
10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
9 + 5*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

589

In [ ]:
n = 10
print( 9 + 5*n + 5*n**2+1,  9 + 5*n + 11*n**2/2  - n/2)

560 604.0


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

589



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 9 + 5*(10) + 5*(10)**2 +  1  = 560 $$

In [ ]:
9 + 5*(10) + 5*(10)**2 +  1

560

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input4.txt
n= 10

560



For the input12.txt
```
200000
200000  199999 ... 2 1
```
$$T = 9 + 5*(200000) + 5*(200000)**2 +1  = 200001000010 $$



In [ ]:
9 + 5*(200000) + 5*(200000)**2 +1

200001000010

Calculate estimated time by instruction and number of instructions by second

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input12.txt


200001000021


In [ ]:
# Estimated time by instruction

306/200001000010


1.5299923499617505e-09

In [ ]:
# Estimated number of instructions by second

306/200001000010

1.5299923499617505e-09

# Estimate the maximun n that can be solved in 1 second

$$5n^2 \times 1.53*10^{-9} s = 1s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.873s
user	0m0.852s
sys	0m0.003s


## Python

In [ ]:
# Verify

## Perform the same analysis for an algorithm in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[0, (n-1)n/2]
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // --------------
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n] upper bound


### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

Test with inputs 1 and 2



Test and time it with input 3

20 stress tests

Count executed instructions in code and verify with inputs 1,2 and 3


Calculate estimated time by instruction and number of instructions by second

### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

Test with inputs 1 and 2



Test and time it with input 3

20 stress tests

Count executed instructions in code and verify with inputs 1,2 and 3


Calculate estimated time by instruction and number of instructions by second

### Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

Test with inputs 1 and 2




Test and time it with input 3

20 stress tests

Count executed instructions in code and verify with inputs 1,2 and 3

Calculate estimated time by instruction and number of instructions by second

### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

Test with inputs 1 and 2




Test and time it with input 3

20 stress tests

Count executed instructions in code and verify with inputs 1,2 and 3

Calculate estimated time by instruction and number of instructions by second